nogo: instantiate type info for generic types when running under Go >=1.18

farhaven opened the pull request:
This mirrors what golang.org/x/tools/go/packages does when loading packages.

Should fix #3211 (and probably #3164 as well).

<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, please read CONTRIBUTING.md and sign the CLA
   first. We cannot review code without a signed CLA.
2. Please file an issue *first*. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
-->

**What type of PR is this?**

> Bug fix

**What does this PR do? Why is it needed?**

This PR adds an initialization that is required for the package loader in nogo to load instantiation information for generic types. This takes a lot of inspiration from what [the package loader in golang.org/x/tools](https://github.com/golang/tools/blob/master/go/packages/packages.go#L952) does.

**Which issues(s) does this PR fix?**

Fixes #3211 and #3164

**Other notes for review**

My company signed the CLA, but it looks like that fact has to percolate through something at Google.

sluongng commented on go/tools/builders/nogo_typeparams_go117.go:
```diff
@@ -0,0 +1,23 @@
+*/
+
+//go:build !go1.18
+// +build !go1.18
```
I think we have been shipping some features in rules_go that are exclusively to 1.18 and are not backward compatible. 🤔 

I don't know if we need to have this file at all 🤔  It would massively simplify this PR down to a single line code change.

farhaven replied:
From the discussion on #3231, I gather that we want to keep the build-tagged files for 1.17 and 1.18, at the very least until 1.19 is released, correct?

linzhp replied:
Yes. We need to stay compatible with Go 1.17 at least until it's deprecated, normally after 1.19 is released.

sluongng commented on go/tools/builders/nogo_typeparams_go118.go:
```diff
@@ -0,0 +1,28 @@
+
+func initInstanceInfo(info *types.Info) {
+	info.Instances = make(map[*ast.Ident]types.Instance)
+}
```
It would be nice if you can turn your small sample repository in the issue into a test in https://github.com/bazelbuild/rules_go/tree/master/tests/core/nogo that is specifically for testing nogo compatibility with Go 1.18's generic.

sluongng reviewed (commented):
@fmeum @achew22 @linzhp I don't think we need to stay compatible with Go 1.17 and below in rules_go? There are some precedent such as https://github.com/bazelbuild/rules_go/pull/3157 where we shipped features exclusively for 1.18

farhaven commented:
I've added a test that's failing for two reasons:

* the Go toolchain version needs to be bumped to >= 1.18. I think there's already a PR for that: #3179. That means that this PR will have to wait until #3179 has been merged. Fine with me :)
* `org_golang_x_tools` needs to be updated, but I don't have (and don't want) the permission to upload packages to mirror.bazel.build

uhthomas commented:
Does this help with https://github.com/bazelbuild/rules_go/issues/3164?

uhthomas commented:
Should the version of x/tools also be updated in this PR? Feels like it would be required for generic associated types.

sluongng commented:
created https://github.com/bazelbuild/bazel-gazelle/issues/1283 to request for a newer Gazelle release

linzhp commented:
Gazelle 0.26.0 was released

sluongng commented:
https://github.com/bazelbuild/rules_go/commit/d6138e4c791948d224600cbf2dce612d3089c1d7 was merged.
I suggest rebase this PR on top of latest origin/master so CI can retry.

farhaven commented:
> [d6138e4](https://github.com/bazelbuild/rules_go/commit/d6138e4c791948d224600cbf2dce612d3089c1d7) was merged. I suggest rebase this PR on top of latest origin/master so CI can retry.

Thanks for the patience, I was on vacation last week and only just got around to rebasing. Let's see how it works.

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+
+func Test(t *testing.T) {
+	cmd := bazel_testing.BazelCmd("query", "//external:org_golang_x_tools", "--output=build")
+	stdout := &bytes.Buffer{}
+	cmd.Stdout = stdout
```
```suggestion
	var stdout bytes.Buffer
	cmd.Stdout = &stdout
```

https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+	t.Log("query:", stdout.String())
+
+	cmd = bazel_testing.BazelCmd("build", "//:src")
+	stderr := &bytes.Buffer{}
+	cmd.Stderr = stderr
```
```suggestion
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
```
https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+		t.Fatal("unexpected error:", err)
+	}
+
+	t.Log("output:", stderr.String())
```
Remove this line to reduce the noise, because stderr is already tested below

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+	t.Log("output:", stderr.String())
+
+	for _, pattern := range []string{
+		"Build completed successfully",
```
This test is brittle: this message is not part of Bazel's public API. When Bazel changes it's output message, this test will fail unnecessarily. It's better to test exit 0, which indicates the build success: https://docs.bazel.build/versions/5.2.0/guide.html#what-exit-code-will-i-get

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+		}
+	}
+	if bytes.Contains(stderr.Bytes(), []byte("panic")) {
+		t.Errorf("%q was printed", "found panic")
```
```suggestion
		t.Errorf("found panic in Bazel output: \n%s", stderr.String())
```

linzhp commented on tests/core/nogo/generics/generics_test.go:
```diff
@@ -0,0 +1,104 @@
+		t.Fatal("failed to run query:", err)
+	}
+
+	t.Log("query:", stdout.String())
```
Either remove this line or move it to one of the error cases below to reduce the noise.

farhaven commented:
I've applied the suggestions, rebased the patch on `master` and squashed the commits down to a single commit to make the history a bit cleaner than "Fix this, fix that, oops, forgot the other thing".
